### PR TITLE
cope with vars with default values

### PR DESCRIPTION
**Pull request by mkeskells**
make Paths "absolute" and "real"
use Path rather than String for filenames
expose flags in toString of elements

### CODE REVIEW

File: analysisPlugin/src/main/scala/org/scalaclean/analysis/ModelSymbolBuilder.scala [hunk 1]
```diff
@@ -82,9 +82,9 @@ trait ModelSymbolBuilder {
           if (gSym.pos == NoPosition) (-1, -1, -1) else (gSym.pos.start, gSym.pos.end, gSym.pos.focus.start)
         val sourceFile =
           if (gSym.sourceFile != null)
-            mungeUnitPath(gSym.sourceFile.file.toPath)
+            gSym.sourceFile.file.toPath.toAbsolutePath.toRealPath()
           else
-            "-"
+            Paths.get("-")
```
Comment (dwijnand) [inline analysisPlugin/src/main/scala/org/scalaclean/analysis/ModelSymbolBuilder.scala:87]: Is Windows cool with "-"?

**Review by dwijnand — APPROVED** [review]
💯 

1) Path > String (though I might eventually propose using RelativePath/AbsolutePath over Path)
2) I have a case of `= _` messing up the dead code removal so I assume this fixes that

File: analysisPlugin/src/main/scala/org/scalaclean/analysis/ScalaCompilerPluginComponent.scala [hunk 2]
```diff
@@ -498,13 +498,33 @@ class ScalaCompilerPluginComponent(val global: Global) extends PluginComponent w
...
+          //its only a var due to for https://github.com/scala/bug/issues/12213 -- see below
+          var symbol = valDef.symbol
           val isVar  = symbol.isVar
           val fields: Option[ModelFields] = valDef.rhs match {
             case Select(qualifier, name) =>
               Option(qualifier.symbol).flatMap(_.attachments.get[ModelFields])
             case _ => None
           }
+          //workaround for https://github.com/scala/bug/issues/12213
+          if (!isVar || !valDef.rhs.isEmpty) valDef.pos.`end` else {
```
Comment (dwijnand) [inline analysisPlugin/src/main/scala/org/scalaclean/analysis/ScalaCompilerPluginComponent.scala:510]: Looks like you can drop `then` part, btw

File: analysisPlugin/src/main/scala/org/scalaclean/analysis/ModelSymbolBuilder.scala (hunk 1 above)
Comment (mkeskells) [reply analysisPlugin/src/main/scala/org/scalaclean/analysis/ModelSymbolBuilder.scala:87]: yes

File: analysisPlugin/src/main/scala/org/scalaclean/analysis/ScalaCompilerPluginComponent.scala (hunk 2 above)
Comment (mkeskells) [reply analysisPlugin/src/main/scala/org/scalaclean/analysis/ScalaCompilerPluginComponent.scala:510]: agreed, fixed

**Comment by mkeskells** [comment]
> 💯
> 
> 1. Path > String (though I might eventually propose using RelativePath/AbsolutePath over Path)
> 2. I have a case of `= _` messing up the dead code removal so I assume this fixes that

1 I think that we can probably drop scalameta dependency completely ( we only use the parser and that Abs/Rel paths I think), but good to enforce that all paths are relative **and real** would be good in some places

2 - yes - see the tests testProjects/deadCodeProject5/src/main/scala/scalaclean/test/deadcode5/DeadCode5VarVal.scala

**Comment by dwijnand** [comment]
Yeah, we'd just keep around copies of what we find useful.

**Pull request merged**